app/assets/javascripts/discourse/app/components/sidebar/edit-navigation-modal-form/tags-form.js

```diff
@@ -25,62 +25,80 @@ export default class extends Component {
   }
 
   async #loadTags() {
-    // This is loading all tags upfront and there is no pagination for it. However, this is what we are doing for the
-    // `/tags` route as well so we have decided to kick this can of worms down the road for now.
+    this.tagsLoading = true;
+
+    const findArgs = {};
+
+    if (this.filter !== "") {
+      findArgs.filter = this.filter;
+    }
+
+    if (this.onlySelected) {
+      findArgs.only_tags = this.selectedTags.join(",");
+    }
+
+    if (this.onlyUnselected) {
+      findArgs.exclude_tags = this.selectedTags.join(",");
+    }
+
     await this.store
-      .findAll("tag")
+      .findAll("listTag", findArgs)
       .then((tags) => {
-        this.tags = tags.content.sort((a, b) => {
-          return a.name.localeCompare(b.name);
-        });
-
         this.tagsLoading = false;
+        this.tags = tags;
       })
       .catch((error) => {
         popupAjaxError(error);
       });
   }
 
-  get filteredTags() {
-    return this.tags.reduce((acc, tag) => {
-      if (this.onlySelected) {
-        if (this.selectedTags.includes(tag.name) && this.#matchesFilter(tag)) {
-          acc.push(tag);
-        }
-      } else if (this.onlyUnselected) {
-        if (!this.selectedTags.includes(tag.name) && this.#matchesFilter(tag)) {
-          acc.push(tag);
-        }
-      } else if (this.#matchesFilter(tag)) {
-        acc.push(tag);
+  @action
+  didInsertTag(element) {
+    const tagName = element.dataset.tagName;
+    const lastTagName = this.tags.content[this.tags.content.length - 1].name;
+
+    if (tagName === lastTagName) {
+      this.observer?.disconnect();
+
+      if (!this.observer) {
+        this.observer = new IntersectionObserver(
+          (entries) => {
+            entries.forEach((entry) => {
+              if (entry.isIntersecting) {
+                this.tags.loadMore();
```

Comment:
Using an `IntersectionObserver` because our `LoadMore` component does not support loading more in modal at all since everything is tightly coupled to the "main" window.

Reply:
Plus I really like this solution because it is so much easier to reason about.